src/timer.rs

```diff
@@ -0,0 +1,70 @@
+trait Timer: Sized {
+    fn start(milliseconds: u32) -> Self;
+    fn has_elapsed(self) -> Option<Self>;
+}
+
+use libtock_core::syscalls;
+
+// see: https://github.com/tock/tock/blob/master/doc/syscalls/00000_alarm.md
+mod command_nr {
+    pub const _IS_DRIVER_AVAILABLE: usize = 0;
+    pub const GET_CLOCK_FREQUENCY: usize = 1;
+    pub const GET_CLOCK_VALUE: usize = 2;
+    pub const _STOP_ALARM: usize = 3;
+    pub const _SET_ALARM: usize = 4;
+}
+const DRIVER_NUMBER: usize = 0x00000;
+
+struct LibtockAlarmTimer {
+    end_tick: usize,
+}
+
+fn wrapping_add_u24(lhs: usize, rhs: usize) -> usize {
+    lhs.wrapping_add(rhs) & 0xffffff
+}
+fn wrapping_sub_u24(lhs: usize, rhs: usize) -> usize {
+    lhs.wrapping_sub(rhs) & 0xffffff
+}
+
+impl Timer for LibtockAlarmTimer {
+    fn start(milliseconds: u32) -> Self {
+        let clock_frequency =
+            syscalls::command(DRIVER_NUMBER, command_nr::GET_CLOCK_FREQUENCY, 0, 0).unwrap_or(0);
+        let start_tick =
+            syscalls::command(DRIVER_NUMBER, command_nr::GET_CLOCK_VALUE, 0, 0).unwrap_or(0);
+        // 32 bits inverted divisor for 1/1000 ( ceil((2^32) / 1000) ), so that (x * INV_DIV) >> 32 ≈ x / 1000
+        const INV_DIV: u64 = 4294968;
+        let delta_tick = ((clock_frequency as u64 * milliseconds as u64 * INV_DIV) >> 32) as usize;
```

Comment:
There's 2 issues here:
- The second multiplication can overflow. And it may overflow in cases where the division would have worked.
- The `INV_DIV` doesn't always work, e.g. for 6100999 it gives one more tick than the division. Such imprecision is probably acceptable given that time is approximative in OpenSK anyway.

I think assuming the frequency is a multiple of 2 and the ticks are on 24 bits is the simplest solution (and this assumption is actually true for all our boards).

Reply:
Indeed there is a slight loss of precision, but I believe it is very limited.

If the second multiplication would overflow, then the next line `assert!(delta_tick < 0x800000);` would as well, as we are checking that the result is within 23 bits, and therefore the result of both multiplications should be within 53 bits (23+32). An easy way to prevent the overflow would be to assert the range after the first multiplication. WDYT?

Reply:
Yes, one solution would be to assert the range after the first multiplication. But then I think this solution is really the simplest:

```rust
// - Assumes frequency is a multiple of 2 (otherwise under-approximates by 1 tick per second).
// - Only unwraps if frequency * milliseconds / 1000 doesn't fit in 23 bits.
let delta_tick = (frequency / 2).checked_mul(milliseconds).unwrap() / 500;
assert!(delta_tick < 0x800000);
```

Reply:
Done.